src/kv/lazy-store.ts

```diff
@@ -0,0 +1,235 @@
+   *
+   * Writes only manipulate the in memory cache (thus values must be persisted to the
+   * base store through a separate process).
+   *
+   * The memory caches size is limited to `cacheSizeLimit`, and values are evicted in an LRU
+   * fashion.  `shouldBePinned` enables exempting some keys from eviction.  This mechanism
+   * should be used to prevent values not persisted to the base store from being lost.
+   * Values whose keys are pinned cannot be evicted, and their sizes are not included in the current
+   * cache size. Once a value is persisted to the base store, it's temp key for which `shouldBePinned`
+   * is `true` can be deleted, and it can be re-put with a key for which `shouldBePinned` is false (making
+   * it subject to cache eviction, which is safe since it is now persisted).
+   *
+   * @param baseStore Store to lazy load and cache values from.
+   * @param cacheSizeLimit Limit in bytes.  Size of values is determined using `getSizeOfValue`.  Keys
+   * do not count towards cache size.  Pinned values (`shouldBePinned`) do not count towards cache
+   * size.
+   * @param shouldBePinned A predicate for determining if a key should be pinned.  Keys
+   * which are not persisted to the underlying base store should be pinned, as this store does
+   * not write through to the base store (it assumes values are persisted to base store via
+   * a separate process).
+   * @param getSizeOfValue Function for measure the size in bytes of a value
+   */
+  constructor(
+    baseStore: Store,
+    cacheSizeLimit: number,
+    shouldBePinned: (key: string) => boolean,
+    getSizeOfValue = defaultGetSizeOfValue,
+  ) {
+    this._baseStore = baseStore;
+    this._cache = new Cache(
+      Math.max(cacheSizeLimit, 0),
+      shouldBePinned,
+      getSizeOfValue,
+    );
+  }
+
+  async read(): Promise<Read> {
+    const release = await this._rwLock.read();
+    return new ReadImpl(this._cache, this._baseStore, release);
+  }
+
+  async withRead<R>(fn: (read: Read) => R | Promise<R>): Promise<R> {
+    const read = await this.read();
+    try {
+      return await fn(read);
+    } finally {
+      read.release();
+    }
+  }
+
+  async write(): Promise<Write> {
+    const release = await this._rwLock.write();
+    return new WriteImpl(
+      this._cache,
+      new ReadImpl(this._cache, this._baseStore, release),
+    );
+  }
+
+  async withWrite<R>(fn: (write: Write) => R | Promise<R>): Promise<R> {
+    const write = await this.write();
+    try {
+      return await fn(write);
+    } finally {
+      write.release();
+    }
+  }
+
+  close(): Promise<void> {
+    return this._baseStore.close();
+  }
+
+  get closed(): boolean {
+    return this._baseStore.closed;
+  }
+}
+
+class ReadImpl implements Read {
+  private readonly _cache: Cache;
+  private readonly _release: () => void;
+  private readonly _base: Store;
+  private _baseRead: Read | undefined;
```

Comment:
This needs an initializer.

I'm surprised TS is not complaining.  https://www.typescriptlang.org/tsconfig#strictPropertyInitialization

> this.address is declared as potentially undefined which means it does not have to be set.

super lame

Reply:
It is important to initialize all properties because adding a property later changes the hidden class (vm implementation detail) which leads to code going from monomorphic to multimorphic.